### PR TITLE
Include signed ip address in TIER2 handshake. #8902 

### PR DESCRIPTION
**Pull request by soonnear**
https://github.com/near/nearcore/issues/8902

The problem being solved
- Prevent man in middle attacks on Tier2 network 

The approach taken
- Pass the signed ip address of the sender 

Remaining steps to fully deploy this solution after this PR is merged
- Fail the handshake only after all nodes on both mainnet and testnet have updated to the latest version of Tier2 handshake that includes signing the ip address. It currently only fails if the signed ip address is present and faulty but doesn't fail if the signed ip address is absent (for backward compatibility temporary migration purposes)

Also needed to run auto-linter to pass buildkite tests
```
cargo fmt -- --emit files 
cargo clippy --fix -- -A clippy::all -W clippy::useless_format -W clippy::clone_on_copy
cargo clippy --fix -- -A clippy::all -W clippy::useless_format -W clippy::redundant_clone
./scripts/formatting --fix
```

### CODE REVIEW

**Comment by soonnear** [comment]
current sanity tests is due to peer signing on its local address of 0.0.0.0 but being advertised as 127.0.0.1, which is only the case for localnet. Will think of a fix for this and update code

**Comment by saketh-are** [comment]
Hey @soonnear, thanks for sharing this. I left some stylistic comments to start, but this looks pretty good overall. Also still need to review tests.

File: chain/network/src/network_protocol/tests.rs
Comment (saketh-are) [inline chain/network/src/network_protocol/tests.rs:181]: This test seems redundant with the more general `serialize_deserialize_protobuf_only` above.

File: chain/network/src/network_protocol/tests.rs
Comment (saketh-are) [inline chain/network/src/network_protocol/tests.rs:211]: This seems like a test of the `sign` function (which accepts an arbitrary sequence of bytes), which should have its own unit tests.

File: chain/network/src/network_protocol/tests.rs
Comment (saketh-are) [inline chain/network/src/network_protocol/tests.rs:224]: Redundant with other tests; also see recommendation not to make a special struct `OwnedIpAddress`.

File: chain/network/src/network_protocol/tests.rs
Comment (saketh-are) [inline chain/network/src/network_protocol/tests.rs:245]: This is a reasonable thing to test, but it is outside the scope of this file. Maybe you could add another test in `connection_pool.rs` testing the happy case (you have two failure cases)?

File: chain/network/src/peer_manager/tests/connection_pool.rs
Comment (saketh-are) [inline chain/network/src/peer_manager/tests/connection_pool.rs:324]: Things should become simpler if you just accept the IpAddr and NodeKey to use directly instead of Options. The tests below can choose to pass either the correct one or an incorrect one.

File: chain/network/src/peer_manager/tests/connection_pool.rs
Comment (saketh-are) [inline chain/network/src/peer_manager/tests/connection_pool.rs:329]: This test is quite long and it would benefit from some visual segmentation for readability. For example, the other tests in this file put an empty line after these first four lines to visually separate the setup of these basic variables from the rest of the test-specific code.

File: chain/network/src/config.rs
Comment (saketh-are) [inline chain/network/src/config.rs:452]: It seems like this changes the definition of a valid config. Have you thought about the implications?

File: chain/network/src/config.rs
Comment (saketh-are) [inline chain/network/src/config.rs:461]: Why would this ever happen? I don't think it makes sense to test this here.

File: chain/network/src/peer/peer_actor.rs
Comment (saketh-are) [inline chain/network/src/peer/peer_actor.rs:572]: See other comments in this function; simplify this to something like "Verify the signed IP Address is valid."

File: chain/network/src/network_protocol/tests.rs
Comment (soonnear) [reply chain/network/src/network_protocol/tests.rs:181]: I preferred adding this fine grained test. It's like saying unit tests are redundant with an end to end integration tests in place. It checks exactly the IpAddr serialize/deserialization whereas the other checks the wrapper.

File: chain/network/src/network_protocol/tests.rs
Comment (soonnear) [reply chain/network/src/network_protocol/tests.rs:211]: The sign function is independent of ip data, this test is specific to checking the sign function works with our serialized approach for ip data that we discussed.

File: chain/network/src/network_protocol/tests.rs
Comment (soonnear) [reply chain/network/src/network_protocol/tests.rs:245]: I spent 4 hours trying to test the happy case, but it just doesnt terminate. I'm not sure if this code allows testing the happy case. I've tested the happy case in other files (peer/communication.rs and integration-tests/network/peer_handshake.rs

File: chain/network/src/peer_manager/tests/connection_pool.rs
Comment (soonnear) [reply chain/network/src/peer_manager/tests/connection_pool.rs:324]: No it doesnt work, cause you need to set up the streams in order to set the IpAddr. so the correct IpAddr cant be passed in.

File: chain/network/src/config.rs
Comment (soonnear) [reply chain/network/src/config.rs:452]: Valid config is simply an internal wrapper around NetworkConfig. It doesnt change how the NetworkConfig is constructed. This passes all existing unit tests, i dont think it affects anything.

Comment (saketh-are) [reply chain/network/src/config.rs:452]: I don't think passing unit tests is sufficient justification. Did you come across any background information on why `node_addr` is an optional field to begin?

File: chain/network/src/config.rs
Comment (saketh-are) [reply chain/network/src/config.rs:461]: Do you have an example of a value which would trigger this check? My understanding is that we can sign an arbitrary sequence of bytes, so I don't understand the purpose of this section of code.

Additionally, take a look at the other checks performed here. They are examining simple boundary conditions on the config (for example, `ideal_connections_lo` should be <= `ideal_connections_hi`). They don't test implementation-specific details but rather basic properties of the config.

File: chain/network/src/network_protocol/proto_conv/net.rs
Comment (soonnear) [reply chain/network/src/network_protocol/proto_conv/net.rs:20]: The other type is SocketAddr, however, we want to sign on IpAddr instead of SocketAddr, so no point sending the SocketAddr information if all we need is just IpAddr. 

This method is necessary because think of it as like, quicksort implements a partition() function within it. But in order to use the partition() function, you first need to extract that code out into a stand-alone function before it can be re-used for other purposes. 

Socket addr contains IpAddr, but I can't use it's protobuf conversion without first extracting that code out of the SocketAddr implementation.

File: chain/network/src/network_protocol/tests.rs
Comment (saketh-are) [reply chain/network/src/network_protocol/tests.rs:181]: `serialize_deserialize_protobuf_only` is precisely a unit test which serializes a message, then deserializes it and checks that all fields match the original message. What kind of future mistake by a developer would this test catch which is not already covered by that test? Notice we don't have tests of this nature for all the other types of primitive fields used in network messages.

File: chain/network/src/network_protocol/tests.rs
Comment (saketh-are) [reply chain/network/src/network_protocol/tests.rs:211]: What does this test catch? The fact that `.octets().to_vec()` produces a `Vec<u8>` is something checked by the compiler. The fact that `sign` will work on an arbitrary sequence of bytes is a test of the `sign` function. I don't think this test justifies its existence here.

File: chain/network/src/config.rs
Comment (soonnear) [reply chain/network/src/config.rs:452]: Good point, looks like its added here to make the listening address configurable for testing purposes
- https://github.com/near/nearcore/pull/8431

Regardless, the listening address is unreliable as the actual communication happens based on the tcp address instead of the listening address. For instance, in pytest's localnet, the listening address is configured as 0.0.0.0, but the actual tcp connecting refers to 127.0.0.1 

updated to this code to verify it passes the pytests
- https://github.com/near/nearcore/commit/16072e5842478de1293e2a7263db04dd055f8d90

will send another commit to clean up and remove any reference to the listening address for signing ip address purposes

File: chain/network/src/config.rs
Comment (soonnear) [reply chain/network/src/config.rs:460]: This discussion is redundant as it's removed from latest code since we'll now use the tcp connection's address instead of configured listening address as mentioned in another comment.

File: chain/network/src/network_protocol/mod.rs
Comment (soonnear) [reply chain/network/src/network_protocol/mod.rs:95]: As discussed offline, it's needed to implement the interface in order to hide the serialization of ip_addr from the user and in order to form the message (SignedOwnedIpAddress) to be sent over the network to contain both required ip address and signature

If i understand correctly, there always needs to be a wrapper class in order to introduce new member methods to existing class such as the examples below
- https://github.com/near/nearcore/blob/7d0b0b4207325f2dab2e77486341a87f5eb111d1/chain/network/src/tcp.rs#L150-L152
- https://github.com/near/nearcore/blob/7d0b0b4207325f2dab2e77486341a87f5eb111d1/chain/network/src/tcp.rs#L60
- https://github.com/near/nearcore/blob/7d0b0b4207325f2dab2e77486341a87f5eb111d1/chain/network/src/tcp.rs#L212

File: chain/network/src/network_protocol/mod.rs
Comment (soonnear) [reply chain/network/src/network_protocol/mod.rs:118]: sure, updated as requested to SignedIpAddress

File: chain/network/src/network_protocol/network.proto
Comment (soonnear) [reply chain/network/src/network_protocol/network.proto:215]: updated as requested

File: chain/network/src/network_protocol/proto_conv/net.rs
Comment (soonnear) [reply chain/network/src/network_protocol/proto_conv/net.rs:1]: As discussed offline, will not be removing the wrapper OwnedIpAddress as it acts as a builder pattern for SignedOwnedIpAddress (renamed to SignedIpAddress) to contain 2 required fields (ip address and signature)

File: chain/network/src/network_protocol/tests.rs
Comment (soonnear) [reply chain/network/src/network_protocol/tests.rs:181]: disagree as discussed offline but this is minor so i have removed.

File: chain/network/src/network_protocol/tests.rs
Comment (soonnear) [reply chain/network/src/network_protocol/tests.rs:213]: this shows how the interface of the new SignedIpAddress object looks like from the client code perspective. Regardless, remove as requested.

File: chain/network/src/network_protocol/tests.rs
Comment (soonnear) [reply chain/network/src/network_protocol/tests.rs:224]: Removed the test but maintain the builder wrapper OwnedIpAddress as discussed offline

File: chain/network/src/network_protocol/tests.rs
Comment (soonnear) [reply chain/network/src/network_protocol/tests.rs:247]: maintain this test since it doesnt go through the network and is mainly specific to protobuf serialization, in which is tested in this specific file.

File: chain/network/src/peer/peer_actor.rs
Comment (soonnear) [reply chain/network/src/peer/peer_actor.rs:115]: Updated to IpAddressMismatch as requested

File: chain/network/src/peer/peer_actor.rs
Comment (soonnear) [reply chain/network/src/peer/peer_actor.rs:576]: Updated as suggested to  // Verify the signed IP address is valid.

File: chain/network/src/peer/tests/communication.rs
Comment (soonnear) [reply chain/network/src/peer/tests/communication.rs:287]: No no, nothing is being specially sent for this test here, this test is to point out that it's currently set to None in test for testing purposes but this test is temporary and will be fully removed once we've mirated.

File: chain/network/src/peer_manager/tests/connection_pool.rs
Comment (soonnear) [reply chain/network/src/peer_manager/tests/connection_pool.rs:333]: Added some white spaces, hopefully that helps.

**Comment by soonnear** [comment]
updated to address first draft review comments, thanks! @saketh-are

File: chain/network/src/network_protocol/tests.rs
Comment (saketh-are) [inline chain/network/src/network_protocol/tests.rs:182]: Don't understand this comment; is it here by mistake?

File: chain/network/src/network_protocol/tests.rs
Comment (saketh-are) [inline chain/network/src/network_protocol/tests.rs:195]: This test is very hard to read, please reconsider the variable names.

Notice that a line such as:
```
    let proto_signed_ip_addr_before_serialize: proto::SignedIpAddr =
        proto::SignedIpAddr::from(&signed_ip_addr_before_serialize);
```

is repeating all kinds of information twice or even more times, which hurts readability.

The name of a variable doesn't need to repeat all possible relevant information, but rather should flow smoothly with the code around it to tell a concise story.

In the context of a test called `serialize_deserialize_signed_ip_address`, simple variable names such as `before` and `after` can be much easier to understand than `signed_ip_addr_before_serialize` and `result_deserialized_actual_signed_ip_addr `.

File: chain/network/src/config.rs
Comment (saketh-are) [inline chain/network/src/config.rs:449]: Looks like this file is not changing meaningfully anymore; can we simply not touch it?

File: chain/network/src/peer/peer_actor.rs
Comment (saketh-are) [inline chain/network/src/peer/peer_actor.rs:293]: What was the address stored previously here and why do we need to change it?

File: core/chain-configs/src/genesis_validate.rs
Comment (saketh-are) [inline core/chain-configs/src/genesis_validate.rs:1]: Changes to this file seem unrelated to this PR?

File: nearcore/src/config_validate.rs
Comment (saketh-are) [inline nearcore/src/config_validate.rs:1]: Changes to this file seem unrelated to this PR?

File: nearcore/src/state_sync.rs
Comment (saketh-are) [inline nearcore/src/state_sync.rs:1]: Changes to this file seem unrelated to this PR?

File: nearcore/src/config_validate.rs
Comment (saketh-are) [inline nearcore/src/config_validate.rs:1]: Changes to this file seem unrelated to this PR

File: nearcore/src/state_sync.rs
Comment (saketh-are) [inline nearcore/src/state_sync.rs:1]: Changes to this file seem unrelated to this PR

File: core/chain-configs/src/genesis_config.rs
Comment (saketh-are) [inline core/chain-configs/src/genesis_config.rs:1]: Changes to this file seem unrelated to this PR

**Comment by saketh-are** [comment]
A better title for this PR would be something like "Include signed ip address in TIER2 handshake."

Also, please add some description which explains what the PR is doing and why. You've linked #8902, the issue being addressed, but have taken a different approach than suggested there ("by implementing a 3-way handshake"). The description should include a few sentences very briefly summarizing:
- The problem being solved
- The approach taken
- Remaining steps to fully deploy this solution after this PR is merged

File: chain/network/src/config.rs
Comment (soonnear) [reply chain/network/src/config.rs:449]: Ah nice catch, reverted. thanks!

File: chain/network/src/network_protocol/tests.rs
Comment (soonnear) [reply chain/network/src/network_protocol/tests.rs:182]: ah my bad, outdated comment. removed. thanks

File: chain/network/src/network_protocol/tests.rs
Comment (soonnear) [reply chain/network/src/network_protocol/tests.rs:195]: ah my bad, i thought i removed this whole test, might have messed up the rebase. Removed.

File: chain/network/src/peer/peer_actor.rs
Comment (soonnear) [reply chain/network/src/peer/peer_actor.rs:293]: it fails the handshake test for localnet. 
Previously stores the listening address, which can be set as 0.0.0.0 and optiona, now will store the actual connection address which will be set as 127.0.0.1 for localnet and is mandatory

File: core/chain-configs/src/genesis_config.rs
Comment (soonnear) [reply core/chain-configs/src/genesis_config.rs:1]: Auto-linter updates these file 
Needed to pass the clippy tests that were added to nayduck, or the PR doesnt pass
- https://github.com/near/nearcore/blob/master/scripts/run_clippy.sh

File: core/chain-configs/src/genesis_validate.rs
Comment (soonnear) [reply core/chain-configs/src/genesis_validate.rs:1]: Same here: 
Auto-linter updates these file
Needed to pass the clippy tests that were added to nayduck, or the PR doesnt pass

https://github.com/near/nearcore/blob/master/scripts/run_clippy.sh

File: nearcore/src/config_validate.rs
Comment (soonnear) [reply nearcore/src/config_validate.rs:1]: Auto-linter updates these file
Needed to pass the clippy tests that were added to nayduck, or the PR doesnt pass

https://github.com/near/nearcore/blob/master/scripts/run_clippy.sh

File: nearcore/src/state_sync.rs
Comment (soonnear) [reply nearcore/src/state_sync.rs:1]: Auto-linter updates these file
Needed to pass the clippy tests that were added to nayduck, or the PR doesnt pass

https://github.com/near/nearcore/blob/master/scripts/run_clippy.sh

**Comment by soonnear** [comment]
note that the unrelated code were added as part of clippy linter auto updates. 
I believe a new rule was added, but it did not update the code in that commit (and that commit did fail the buildkite), so i've to update here or it wont pass the nayduck clippy tests. 
- https://github.com/near/nearcore/commit/132b52a58b82f9bc8b5dea80b74d8dd352ddd296

**Comment by soonnear** [comment]
Thanks! Updated both the PR title and description as suggested. 

Also updated code based on 2nd revision comments, thanks! @saketh-are

File: chain/network/src/peer/peer_actor.rs
Comment (saketh-are) [reply chain/network/src/peer/peer_actor.rs:293]: Ok, I think I understand what's going on from some of your other comments. `network_state.config.node_addr` has the ip address on which the node listens for new connections. When that listener does accept a connection and create a new stream, it is more correct/direct to just take the address for the stream from the stream object rather than assume it will be identical to the one for the listener and indirectly get the address through the network config.

Can you remove or modify this comment? It is not clear to me what is being said (for example we are not "updating" anything here, it is the spawn logic and everything is being set for the first time).

File: chain/network/src/testonly/stream.rs
Comment (saketh-are) [inline chain/network/src/testonly/stream.rs:10]: I think you can make this `pub (crate)`. In general, aim to limit visibility to the lowest level needed.

File: chain/network/src/network_protocol/tests.rs
Comment (saketh-are) [reply chain/network/src/network_protocol/tests.rs]: Resolving conversation as it seems this test was removed.

File: chain/network/src/peer_manager/tests/connection_pool.rs
Comment (saketh-are) [reply chain/network/src/peer_manager/tests/connection_pool.rs:333]: Not sure which test you changed but `test_backward_compatible_handshake_without_signed_ip_address` still looks to have this issue.

**Comment by saketh-are** [comment]
Thanks @soonnear, this is looking good. I've left just a few final comments; this should be ready to merge once you've taken a look.

File: chain/network/src/testonly/stream.rs
Comment (soonnear) [reply chain/network/src/testonly/stream.rs:10]: Cool, didn't know the differences but just learnt it here
- https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/reference/visibility-and-privacy.html#pubin-path-pubcrate-pubsuper-and-pubself

updated to `pub (crate)` as suggested.

File: chain/network/src/peer/peer_actor.rs
Comment (soonnear) [reply chain/network/src/peer/peer_actor.rs:293]: Yes, you understood it perfectly. Took me some time diving into the code to figure out the differences between both socket addresses as well. 
I removed the comment instead of updating the comment.

File: chain/network/src/peer_manager/tests/connection_pool.rs
Comment (soonnear) [reply chain/network/src/peer_manager/tests/connection_pool.rs:333]: Ah lol, the test `test_backward_compatible_handshake_without_signed_ip_address` is in a different file, this comment is pointing towards connection_pool.rs whereas that test is in the file communication.rs 

I previously updated the `test_signed_ip_address` which is being referred to by this code review comment. 

Updated the test as suggested, however, do note that this backward compatibility test is just temporary and will be removed once all nodes have fully migrated to signing ip address.

File: chain/network/src/network_protocol/mod.rs
Comment (soonnear) [reply chain/network/src/network_protocol/mod.rs:95]: Regarding 1. good point, I thought about it before but somehow forgot it. I guess I preferred the other approach at the time as I could refer to protobuf serialization of OwnedAccount. Updated to SignedIpAddress with constructors as suggested. 

Ignoring 2. since it's redundant thanks to 1.

File: chain/network/src/network_protocol/mod.rs
Comment (saketh-are) [inline chain/network/src/network_protocol/mod.rs:93]: Can we revise this comment to say what's really going on here? As mentioned previously ip addresses aren't really "owned." I would suggest something like:
```
// Signed ip address included as part of the Handshake message.
// Used to authenticate identity of connected peers.
// Necessary because peer discovery happens via unauthenticated PeerInfos.
```

File: chain/network/src/network_protocol/mod.rs
Comment (saketh-are) [inline chain/network/src/network_protocol/mod.rs:118]: Why do we need the separate helper function? Can we just inline it here?

**Review by saketh-are — APPROVED** [review]
Thanks, LGTM

**Review by saketh-are — APPROVED** [review]
Thanks, LGTM

**Pull request merged**
